src/main/java/com/uber/cadence/workflow/Saga.java

```diff
@@ -0,0 +1,37 @@
+package com.uber.cadence.workflow;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+
+public final class Saga {
+  private final Deque<Functions.Proc> compensations = new ArrayDeque<>();
+  private final boolean parallelCompensation;
+
+  public Saga(boolean parallelCompensation) {
+    this.parallelCompensation = parallelCompensation;
+  }
+
+  public void addCompensation(Functions.Proc func) {
+    compensations.addFirst(func);
+  }
+
+
+  public void compensate() {
```

Comment:
What should we do if compensation fails?

Reply:
Have an option to either ignore errors or fail.

Reply:
With both `parallelCompensation` and `continueWithError` options, compensation behavior looks a bit implicit:

| parallelCompensation  | continueWithError | real behavior in case of failure |
| ------------- | ------------- |--------|
| false  | false  | as expected |
| false  | true | as expected |
| true  | false  | efficiently ignored? |
| true  | true | as expected |

Reply:
I think it is more about reporting if error happened. When continueWithError is false all errors are ignored.